services/federator/test/unit/Test/Federator/Validation.hs

```diff
@@ -77,26 +101,125 @@ validateDomainAllowListFailSemantic =
+      res :: Either InwardError Domain <-
+        Polysemy.runError
+          . mockDiscoveryTrivial
+          . Polysemy.runReader settings
+          $ validateDomain (Just exampleCert) ("invalid//.><-semantic-&@-domain" :: Text)
+      case res of
+        Left (InwardError IInvalidDomain _) -> pure ()
+        x -> embed $ assertFailure $ "expected IInvalidDomain error, got " <> show x
 
 validateDomainAllowListFail :: TestTree
 validateDomainAllowListFail =
   testCase "allow list validation" $
     runM . evalMock @Remote @IO $ do
       let settings = settingsWithAllowList [Domain "only.other.domain"]
-      res :: Either InwardError Domain <- Polysemy.runError . Polysemy.runReader settings $ validateDomain ("hello.world" :: Text)
-      embed $ assertEqual "allow list:" (Left IFederationDeniedByRemote) (mapLeft inwardErrorType res)
+      exampleCert <- embed $ BS.readFile "test/resources/unit/localhost.example.com.pem"
+      res :: Either InwardError Domain <-
+        Polysemy.runError
+          . mockDiscoveryTrivial
+          . Polysemy.runReader settings
+          $ validateDomain (Just exampleCert) ("localhost.example.com" :: Text)
+      case res of
+        Left (InwardError IFederationDeniedByRemote _) -> pure ()
+        x -> embed $ assertFailure $ "expected IFederationDeniedByRemote error, got " <> show x
 
 validateDomainAllowListSuccess :: TestTree
 validateDomainAllowListSuccess =
   testCase "should give parsed domain if in the allow list" $
     -- removing evalMock @Remote doesn't seem to work, but why?
     runM . evalMock @Remote @IO $ do
-      let domain = Domain "hello.world"
+      let domain = Domain "localhost.example.com"
       let settings = settingsWithAllowList [domain]
-      res :: Either InwardError Domain <- Polysemy.runError . Polysemy.runReader settings $ validateDomain ("hello.world" :: Text)
-      embed $ assertEqual "validateDomain should give 'hello.world' as domain" (Right domain) res
+      exampleCert <- embed $ BS.readFile "test/resources/unit/localhost.example.com.pem"
+      res :: Either InwardError Domain <-
+        Polysemy.runError
+          . mockDiscoveryTrivial
+          . Polysemy.runReader settings
+          $ validateDomain (Just exampleCert) (domainText domain)
+      embed $ assertEqual "validateDomain should give 'localhost.example.com' as domain" (Right domain) res
+
+validateDomainCertMissing :: TestTree
+validateDomainCertMissing =
+  testCase "should fail if no client certificate is provided" $
+    runM . evalMock @Remote @IO $ do
+      res :: Either InwardError Domain <-
+        Polysemy.runError
+          . mockDiscoveryTrivial
+          . Polysemy.runReader noClientCertSettings
+          $ validateDomain Nothing "foo.example.com"
+      case res of
+        Left (InwardError IInvalidDomain _) -> pure ()
+        x -> embed $ assertFailure $ "expected IInvalidDomain error, got " <> show x
```

Comment:
Seems weird that everything is `IInvalidDomain`, I think a better name would be `IAuthenticationFailed`, I am going to try and rename this.